php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+            		'   AND ds.Override = false '.
+            		' GROUP BY SkillID '.
+            ' ) AS skillRatings ',
+            [
+                DemonstrationSkill::$tableName,
+                Demonstration::$tableName,
+                join(', ', $validSkillIds),
+                $this->Level,
+                $this->StudentID
+            ]
+        );
+
+        // check if each skill has been rated atleast once for the current level
+        if ((int)$ratedSkills !== count($validSkills)) {
+            return null;
+        }
+
+        // get the first rating (by date) for each skill
+        $demonstrationSkillRatings = DB::allValues(
+            'skillRatings',
+            'SELECT ds.DemonstratedLevel as skillRatings'.
+        	'  FROM `%1$s` ds '.
+    		'  JOIN `%2$s` d '.
+    		'    ON d.ID = ds.DemonstrationID '.
+    		' INNER JOIN ( '.
+                    'SELECT MIN(demoskill.Created) as Created, SkillID '.
+    		        '  FROM `%1$s` demoskill '.
+    		        '  JOIN `%2$s` demo '.
+    		        '    ON demo.ID = demoskill.DemonstrationID '.
+    		        ' WHERE TargetLevel = %3$u '.
+                    '   AND DemonstratedLevel > 0 '.
+                    '   AND Override = false '.
+                    '   AND demoskill.SkillID IN (%4$s) '.
+                    '   AND demo.StudentID = %5$u '.
+                    ' GROUP BY SkillID '.
+    		'  ) ds2 '.
+    		'    ON ds2.SkillID = ds.SkillID '.
+    		'   AND ds2.Created = ds.Created '.
+    		' WHERE ds.SkillID IN (%4$s) '.
+    		'   AND ds.TargetLevel = %3$u '.
+    		'   AND ds.DemonstratedLevel > 0 '.
+    		'   AND d.StudentID = %5$u '.
+    		'   AND ds.Override = false ',
+            [
+                DemonstrationSkill::$tableName,
+                Demonstration::$tableName,
+                $this->Level,
+                join(', ', $validSkillIds),
+                $this->StudentID
+            ]
+        );
+
+        $this->BaselineRating = array_sum($demonstrationSkillRatings) / count($validSkills);
+
+        if ($autoSave === true) {
+            $this->save();
+        }
+    }
+
+    private $demonstrationData;
+    public function getDemonstrationData()
+    {
+        if (!empty($this->demonstrationData)) {
+            return $this->demonstrationData;
+        }
+
+        try {
+            $data = DB::arrayTable(
+                'SkillID',
+                'SELECT '.
+                '    DemonstrationSkill.*, '.
+                '    Demonstration.Demonstrated AS DemonstrationDate '.
+                '  FROM `%s` DemonstrationSkill '.
+                '  JOIN (SELECT ID, Demonstrated FROM `%s` WHERE StudentID = %u) Demonstration '.
+                '    ON Demonstration.ID = DemonstrationSkill.DemonstrationID '.
+                ' WHERE DemonstrationSkill.SkillID IN (%s) '.
+                '   AND DemonstrationSkill.TargetLevel = %u '.
+                ' ORDER BY SkillID, DemonstrationDate',
```

Comment:
order by DemonstrationID after DemonstartionDate so that multiple demonstrations on the same date get sorted consistently by order entered